extensions/arc/src/ui/dialogs/connectControllerDialog.ts

```diff
@@ -5,101 +5,142 @@
-				.withProperties<azdata.InputBoxProperties>({
-					value: controllerInfo?.username
-				}).component();
-			this.passwordInputBox = this.modelBuilder.inputBox()
-				.withProperties<azdata.InputBoxProperties>({
-					inputType: 'password',
-					value: password
-				})
-				.component();
-			this.rememberPwCheckBox = this.modelBuilder.checkBox()
-				.withProperties<azdata.CheckBoxProperties>({
-					label: loc.rememberPassword,
-					checked: controllerInfo?.rememberPassword
-				}).component();
+			this.initializeFields(controllerInfo, password);
 
 			let formModel = this.modelBuilder.formContainer()
 				.withFormItems([{
-					components: [
-						{
-							component: this.urlInputBox,
-							title: loc.controllerUrl,
-							required: true
-						}, {
-							component: this.nameInputBox,
-							title: loc.controllerName,
-							required: false
-						}, {
-							component: this.usernameInputBox,
-							title: loc.username,
-							required: true
-						}, {
-							component: this.passwordInputBox,
-							title: loc.password,
-							required: true
-						}, {
-							component: this.rememberPwCheckBox,
-							title: ''
-						}
-					],
+					components: this.getComponents(),
 					title: ''
 				}]).withLayout({ width: '100%' }).component();
 			await view.initializeModel(formModel);
-			this.urlInputBox.focus();
+			await this.fieldToFocusOn().focus();
+			this.readonlyFields().forEach(f => f.readOnly = true);
 			this.initialized = true;
 		});
 
-		dialog.registerCloseValidator(async () => await this.validate());
-		dialog.okButton.label = loc.connect;
-		dialog.cancelButton.label = loc.cancel;
-		azdata.window.openDialog(dialog);
-		return dialog;
+		this.dialog.registerCloseValidator(async () => await this.validate());
+		this.dialog.okButton.label = loc.connect;
+		this.dialog.cancelButton.label = loc.cancel;
+		azdata.window.openDialog(this.dialog);
+		return this.dialog;
+	}
+
+	public abstract async validate(): Promise<boolean>;
+
+	private handleCancel(): void {
+		this._completionPromise.resolve(undefined);
+	}
+
+	public waitForClose(): Promise<ConnectToControllerDialogModel | undefined> {
+		return this._completionPromise.promise;
+	}
+}
+
+export class ConnectToControllerDialog extends ControllerDialogBase {
+	protected rememberPwCheckBox!: azdata.CheckBoxComponent;
+
+	protected fieldToFocusOn = () => this.urlInputBox;
+
+	protected getComponents() {
+		return [
+			...super.getComponents(),
+			{
+				component: this.rememberPwCheckBox,
+				title: ''
+			}];
+	}
+
+	protected initializeFields(controllerInfo: ControllerInfo | undefined, password: string | undefined) {
+		super.initializeFields(controllerInfo, password);
+		this.rememberPwCheckBox = this.modelBuilder.checkBox()
+			.withProperties<azdata.CheckBoxProperties>({
+				label: loc.rememberPassword,
+				checked: controllerInfo?.rememberPassword
+			}).component();
+	}
+
+	constructor(protected _treeDataProvider: AzureArcTreeDataProvider) {
```

Comment:
Protected vars shouldn't have _

Reply:
And anyways - you already have the treeDataProvider defined in the base class so you can just remove the protected modifier here and pass it to the super constructor. (still remove the _)